### PR TITLE
[13.0][MIG] excel_import_export, excel_import_export_demo

### PR DESCRIPTION
**Pull request by kittiu**
Standard Migration + Enhancement (easy report option, add contextual action)

Ready for review.

### CODE REVIEW

**Comment by kittiu** [comment]
cc @pafnow @tarbib

**Comment by kittiu** [comment]
cc @kittiu https://github.com/OCA/server-tools/pull/1899#pullrequestreview-487653572

File: excel_import_export_demo/import_export_purchase_order/templates.xml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+<!--
+  Copyright 2019 Ecosoft Co., Ltd.
+  License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).-->
+<odoo>
+    <record id="purchase_order_xlsx_template" model="xlsx.template">
+        <field name="res_model">purchase.order</field>
+        <field name="fname">purchase_order.xlsx</field>
+        <field name="name">Purchase Order Template (import/export)</field>
+        <field name="description">Sample Purchase Order Template for testing</field>
+        <field name="import_action_id" eval="ref('action_purchase_oder_export_xlsx')" />
```
Comment (newtratip) [inline excel_import_export_demo/import_export_purchase_order/templates.xml]:
```suggestion
        <field name="import_action_id" eval="ref('action_purchase_oder_import_xlsx')" />
```

File: excel_import_export_demo/import_export_purchase_order/templates.xml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        <field name="export_action_id" eval="ref('action_purchase_oder_import_xlsx')" />
```
Comment (newtratip) [inline excel_import_export_demo/import_export_purchase_order/templates.xml]:
```suggestion
        <field name="export_action_id" eval="ref('action_purchase_oder_export_xlsx')" />
```

File: excel_import_export_demo/import_export_sale_order/templates.xml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+<!--
+  Copyright 2019 Ecosoft Co., Ltd.
+  License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).-->
+<odoo>
+    <record id="sale_order_xlsx_template" model="xlsx.template">
+        <field name="res_model">sale.order</field>
+        <field name="fname">sale_order.xlsx</field>
+        <field name="name">Sale Order Template (import/export)</field>
+        <field name="description">Sample Sales Order Template for testing</field>
+        <field name="import_action_id" eval="ref('action_sale_oder_export_xlsx')" />
```
Comment (newtratip) [inline excel_import_export_demo/import_export_sale_order/templates.xml:11]:
```suggestion
        <field name="import_action_id" eval="ref('action_sale_oder_import_xlsx')" />
```

File: excel_import_export_demo/import_export_sale_order/templates.xml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        <field name="export_action_id" eval="ref('action_sale_oder_import_xlsx')" />
```
Comment (newtratip) [inline excel_import_export_demo/import_export_sale_order/templates.xml:12]:
```suggestion
        <field name="export_action_id" eval="ref('action_sale_oder_export_xlsx')" />
```

**Review by newtratip — COMMENTED** [review]
Some review on excel_import_export_demo

File: excel_import_export_demo/import_export_sale_order/templates.xml (hunk 3 above)
Comment (kittiu) [reply excel_import_export_demo/import_export_sale_order/templates.xml:11]: Done

File: excel_import_export_demo/import_export_sale_order/templates.xml (hunk 4 above)
Comment (kittiu) [reply excel_import_export_demo/import_export_sale_order/templates.xml:12]: Done.

File: excel_import_export_demo/import_export_purchase_order/templates.xml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+<!--
+  Copyright 2019 Ecosoft Co., Ltd.
+  License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).-->
+<odoo>
+    <record id="purchase_order_xlsx_template" model="xlsx.template">
+        <field name="res_model">purchase.order</field>
+        <field name="fname">purchase_order.xlsx</field>
+        <field name="name">Purchase Order Template (import/export)</field>
+        <field name="description">Sample Purchase Order Template for testing</field>
+        <field name="import_action_id" eval="ref('action_purchase_order_import_xlsx')" />
```
Comment (alessandro-fiorino) [inline excel_import_export_demo/import_export_purchase_order/templates.xml]: <field
            name="import_action_id"
            eval="ref('action_purchase_order_import_xlsx')"
        />

File: excel_import_export_demo/import_export_purchase_order/templates.xml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        <field name="export_action_id" eval="ref('action_purchase_order_export_xlsx')" />
```
Comment (alessandro-fiorino) [inline excel_import_export_demo/import_export_purchase_order/templates.xml]: <field
            name="export_action_id"
            eval="ref('action_purchase_order_export_xlsx')"
        />

**Review by alessandro-fiorino — CHANGES_REQUESTED** [review]

File: excel_import_export/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        "security/ir.model.access.csv",
+        "wizard/export_xlsx_wizard.xml",
+        "wizard/import_xlsx_wizard.xml",
+        "wizard/report_xlsx_wizard.xml",
+        "views/xlsx_template_view.xml",
+        "views/xlsx_report.xml",
+        "views/webclient_templates.xml",
+    ],
+    "installable": True,
+    "development_status": "beta",
```
Comment (alessandro-fiorino) [inline excel_import_export/__manifest__.py]: It seems that "development_status" : beta is not liked by Travis..

**Review by alessandro-fiorino — CHANGES_REQUESTED** [review]

**Comment by alessandro-fiorino** [comment]
How can code coverage be improved ? Maybe adding some test functions to the common.py unit ?

File: excel_import_export_demo/import_sale_orders/templates.xml [hunk 8]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!--
+  Copyright 2019 Ecosoft Co., Ltd.
+  License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).-->
+<odoo>
+    <record id="import_sale_order_xlsx_template" model="xlsx.template">
+        <field name="res_model">sale.order</field>
+        <field name="fname">import_sale_order.xlsx</field>
+        <field name="name">Import Sale Order Template</field>
+        <field name="description">Sample Import Sales Order Tempalte for testing</field>
```
Comment (Siraparn) [inline excel_import_export_demo/import_sale_orders/templates.xml:10]: Sample Import Sales Order Template for testing

File: excel_import_export_demo/tests/test_common.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        categ_ids = cls.env["res.partner.category"].search([]).ids
+        cls.partner = cls.env["res.partner"].create(
+            {"name": "Test Partner", "category_id": [(6, 0, categ_ids)]}
+        )
+        # Create a Product
+        user_type_income = cls.env.ref("account.data_account_type_direct_costs")
+        cls.account_income_product = cls.env["account.account"].create(
+            {
+                "code": "INCOME_PROD111",
+                "name": "Icome - Test Account",
```
Comment (Siraparn) [inline excel_import_export_demo/tests/test_common.py]: Income - Test Account

File: excel_import_export_demo/__manifest__.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+
+{
+    "name": "Excel Import/Export/Report Demo",
+    "version": "13.0.1.0.0",
+    "author": "Ecosoft,Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "website": "https://github.com/OCA/server-tools/",
+    "category": "Tools",
+    "depends": ["excel_import_export", "sale_management", "purchase"],
```
Comment (alessandro-fiorino) [inline excel_import_export_demo/__manifest__.py]: Add "crm" as dependancy for the tests to correct this runbot errors:

2020-11-02 20:48:29,363 159 WARNING openerp_test odoo.fields: Field report.crm.lead.results with unknown comodel_name 'crm.lead'
2020-11-02 20:48:29,363 159 WARNING openerp_test odoo.fields: Field report.crm.lead.revenue_by_country with unknown comodel_name 'crm.lead'
2020-11-02 20:48:29,363 159 WARNING openerp_test odoo.fields: Field report.crm.lead.revenue_by_team with unknown comodel_name 'crm.lead'

**Review by alessandro-fiorino — CHANGES_REQUESTED** [review]
Add "crm" as dependancy to the demo to avoid runbot errors

File: excel_import_export_demo/__manifest__.py (hunk 10 above)
Comment (alessandro-fiorino) [inline excel_import_export_demo/__manifest__.py]: `"depends": ["excel_import_export", "sale_management", "purchase", "crm"],`

**Review by alessandro-fiorino — CHANGES_REQUESTED** [review]

File: excel_import_export_demo/import_sale_orders/templates.xml (hunk 8 above)
Comment (kittiu) [reply excel_import_export_demo/import_sale_orders/templates.xml:10]: Fix

File: excel_import_export_demo/tests/test_common.py (hunk 9 above)
Comment (kittiu) [reply excel_import_export_demo/tests/test_common.py]: Fix

File: excel_import_export_demo/__manifest__.py (hunk 10 above)
Comment (kittiu) [reply excel_import_export_demo/__manifest__.py]: Fixed

File: excel_import_export/__manifest__.py (hunk 7 above)
Comment (kittiu) [reply excel_import_export/__manifest__.py]: Fixed thanks.

**Comment by kittiu** [comment]
> How can code coverage be improved ? Maybe adding some test functions to the common.py unit ?

May we keep test at this level at the moment. Still Beta :)

**Review by alessandro-fiorino — CHANGES_REQUESTED** [review]
Travis prefer single lines on a couple of fields declaration

**Comment by kittiu** [comment]
@alessandro-fiorino 
@Siraparn 
@newtratip 
If this is ok, can you help approve this module?

Note: this module is still on Alpha, and so it might require more enhancement in the future as use cases grows. But we have used it in many projects and use cases.

**Review by newtratip — APPROVED** [review]
It's work as well.

**Review by alessandro-fiorino — APPROVED** [review]
Well done!

**Comment by kittiu** [comment]
Thanks all !!
@JordiBForgeFlow can we merge? I think it is ready to move.

**Comment by Thanh-Tu7785** [comment]
Hi all,
Can we have the final release for this module? We want to apply this.
Thanks

**Comment by kittiu** [comment]
@OCA/tools-maintainers can this get merged? Or anything else todo.
Thanks!

**Review by Saran440 — APPROVED** [review]
Functional Test :+1:

**Comment by kittiu** [comment]
@dreispt I just rebased, can you try again please?

**Comment by kittiu** [comment]
> @dreispt I just rebased, can you try again please?

@dreispt would be great if this can get merged. This module is pending here for quite some time, we did used it a lot in projects.

Thanks!

File: excel_import_export/readme/USAGE.rst [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    />
+
+By using report action, Odoo will find template using combination of model and name, then do the export for the underlining record.
+Please see example in excel_import_export_demo/report_action, which shows,
+
+1. Print excel from an active sale.order
+2. Run partner list report based on search criteria.
+
+Easy Reporting Option
+---------------------
```
Comment (JordiMForgeFlow) [inline excel_import_export/readme/USAGE.rst]:
```suggestion
~~~~~~~~~~~~
```

File: excel_import_export/readme/USAGE.rst [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+For reporting, also call `export_xlsx(...)` but through following method
+
+- ``self.env['xslx.report'].report_xlsx(...)``
+
+After install this module, go to Settings > Excel Import/Export > XLSX Templates, this is where the key component located.
+
+As this module provide tools, it is best to explain as use cases. For example use cases, please install **excel_import_export_demo**
+
+Use Cases
+---------
```
Comment (JordiMForgeFlow) [inline excel_import_export/readme/USAGE.rst]:
```suggestion
~~~~~
```

File: excel_import_export/readme/USAGE.rst [hunk 13]
```diff
@@ -0,0 +1,88 @@
+Concepts
+--------
```
Comment (JordiMForgeFlow) [inline excel_import_export/readme/USAGE.rst]:
```suggestion
~~~~~
```

**Comment by JordiMForgeFlow** [comment]
I faced the same issue in another PR, I think that the above suggestion changes should fix it.

**Comment by JordiMForgeFlow** [comment]
Hi @kittiu and @dreispt,

Could we try to merge this? :)

**Pull request merged**